runtime/sema/type.go

```diff
@@ -7982,12 +8147,15 @@ type IntersectionType struct {
 	memberResolvers              map[string]MemberResolver
 	memberResolversOnce          sync.Once
 	supportedEntitlements        *EntitlementOrderedSet
+	// Deprecated
+	LegacyType Type
 }
 
 var _ Type = &IntersectionType{}
 
-func NewIntersectionType(memoryGauge common.MemoryGauge, types []*InterfaceType) *IntersectionType {
-	if len(types) == 0 {
+// TODO: remove `legacyType` once all uses of it are removed
+func NewIntersectionType(memoryGauge common.MemoryGauge, legacyType Type, types []*InterfaceType) *IntersectionType {
```

Comment:
Personally, I prefer to have the constructor not have a legacy type, and to require setting it manually in the few times it's actually used. IMO this makes it clear that this isn't something to be used in most code, but ultimately it's just a matter of preference I guess.

Reply:
The legacy type is used to check the validity of the constructed type (`len(types) == 0 && legacyType == nil` is invalid). I'd like to keep the check in place, as it helped us already identify some cases we did not handle properly